cedar-policy-validator/src/typecheck.rs

```diff
@@ -1087,9 +1179,23 @@ impl<'a> Typechecker<'a> {
             ),
             // Template Slots, always has to be an entity.
             ExprKind::Slot(slotid) => TypecheckAnswer::success(
-                ExprBuilder::with_data(Some(Type::any_entity_reference()))
-                    .with_same_source_info(e)
-                    .slot(*slotid),
+                ExprBuilder::with_data(Some(if slotid.is_principal() {
+                    request_env
+                        .principal_slot
+                        .clone()
+                        .map(|ety| Type::possibly_unspecified_entity_reference(ety))
+                        .unwrap_or(Type::any_entity_reference())
+                } else if slotid.is_resource() {
+                    request_env
+                        .resource_slot
+                        .clone()
+                        .map(|ety| Type::possibly_unspecified_entity_reference(ety))
+                        .unwrap_or(Type::any_entity_reference())
+                } else {
+                    Type::any_entity_reference()
```

Comment:
Should these `unwrap` and `else` cases be errors? Why is it ok to return `any_entity_reference`?

Reply:
If I see some other slot that I didn't account for in the query env, it's still safe to call it `AnyEntity`. This could be a panic, because it shouldn't happen currently, but this behavior gives a sound fall-back if it does occur.